File: schemas/auth-rbac/drizzle/mfa_recovery_codes.ts
// mfa_recovery_codes: Backup codes for when a user loses access to their MFA device.
// Each code is a separate row for individual consumption tracking.
// See README.md for full design rationale and field documentation.

import { index, pgTable, text, timestamp, uuid } from "drizzle-orm/pg-core";

import { users } from "./users";

export const mfaRecoveryCodes = pgTable(
  "mfa_recovery_codes",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),

    // Hashed recovery code. The plaintext is shown once at generation, never again.
    codeHash: text("code_hash").notNull(),
    // NULL = available. Set when consumed. A used code cannot be reused.
    usedAt: timestamp("used_at", { withTimezone: true }),

    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    // "Get all recovery codes for this user" (to check how many are left).
    index("idx_mfa_recovery_codes_user_id").on(table.userId),
  ],
);

File: schemas/auth-rbac/spacetimedb/mfa_recovery_codes.rs
// mfa_recovery_codes: Backup codes for when a user loses access to their MFA device.
// Generated as a batch (e.g., 10 codes) when MFA is first enabled. Each code is a separate row.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = mfa_recovery_codes, public)]
pub struct MfaRecoveryCode {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub user_id: String, // FK → users.id (cascade delete). Index to count remaining codes.

    pub code_hash: String, // Hashed recovery code. Plaintext shown once at generation, never again.

    pub used_at: Option<Timestamp>, // Null = available. Set when consumed. A used code cannot be reused.

    pub created_at: Timestamp,
}
